core/primitives/src/utils.rs

```diff
@@ -114,15 +137,21 @@ impl KeyForAccount {
     }
 }
 
-#[derive(derive_more::AsRef, derive_more::Into)]
-#[as_ref(forward)]
+#[derive(derive_more::Into)]
 pub struct KeyForAccessKey(Vec<u8>);
 
+impl_trie_key!(KeyForAccessKey);
+
 impl KeyForAccessKey {
     fn estimate_prefix_len(account_id: &AccountId) -> usize {
         KeyForColumnAccountId::estimate_len(col::ACCESS_KEY, account_id) + col::ACCESS_KEY.len()
     }
 
+    /// This is not safe and should only be used internally for iterating over access keys for reading.
+    pub fn from_raw_key(key: &[u8]) -> Self {
+        Self(key.to_vec())
```

Comment:
I believe we can avoid this if we refactor the only place where this is needed: `NightShadeRuntime::view_access_keys`. Given that we create a `KeyForAccessKey` with `get_prefix` there, filtering by such a key should yield records with a key that is of `KeyForAccessKey` type.

Strictly speaking, this method should be marked as `unsafe` from the API point of view as it requires some external contract to be upheld (just like any `*_unchecked` method in Rust std library). I understand your hesitation to introduce unsafe code, I have been there just a few days ago, and ended up sourcing the iterator handling into the `KeyForStateChanges`:

https://github.com/nearprotocol/nearcore/blob/62f00b211915790d06e822da866fb01ceb6cdc89/core/store/src/trie/mod.rs#L622-L639

I am not 100% happy with it as I had to pollute the `KeyFor*` implementation with something a bit orthogonal, but I did not need to "hide" or introduce unsafe `from_raw_key` method.

Reply:
I think it's okay to have it here until we remove iterators completely. Once we remove iterators and replace them with appropriate implementations based on unordered key-value store, we can remove `from_raw_key` methods.